pyanaconda/modules/base.py

```diff
@@ -0,0 +1,61 @@
+# base.py
+# Anaconda DBUS module base.
+#
+# Copyright (C) 2017 Red Hat, Inc.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+import gi
+gi.require_version("GLib", "2.0")
+from gi.repository import GLib
+from pydbus import SessionBus
+
+from pyanaconda import constants
+from pyanaconda.dbus import get_bus
+from pyanaconda.dbus.interface import dbus_interface
+
+from pyanaconda import anaconda_logging
+log = anaconda_logging.get_dbus_module_logger(__name__)
+
+@dbus_interface(constants.DBUS_MODULE_BASE)
+class BaseModule(object):
+    """A common base for Anaconda DBUS modules.
+
+    This class also basically defines the common DBUS API
+    of Anaconda DBUS modules.
+    """
+
+    module_name = None
+
+    def __init__(self):
+        self._loop = GLib.MainLoop()
+        self._bus = get_bus()
+        self._dbus = self._bus.get('org.freedesktop.DBus', '/org/freedesktop/DBus')
+
+    @property
+    def full_module_name(self):
+        return "%s.%s" % (constants.DBUS_MODULE_NAMESPACE, self.module_name)
+
+    def publish_module(self):
+        """Publish the module on DBUS and start its main loop."""
+        self._bus.publish(self.full_module_name, self)
+        log.debug("module %s has been published", self.full_module_name)
+        self._loop.run()
+
+    def Quit(self):
+        """Shut the module down."""
+        log.debug("module quiting: %s", self.full_module_name)
+        self._loop.quit()
```

Comment:
Just a note:
This quit is not quite clean (it is not supposed to be at this point I guess) - it will (might) cause a bus error (probably just a warning) about not getting the reply for the Quit call. To prevent it we'd either have to give the module some time to send the reply before stopping the loop or not require reply for the method (NO_REPLY_EXPECTED, but I doubt pydbus supports this). In any case it is not a big deal now.